nexus/db-queries/src/db/datastore/allowed_source_ips.rs

```diff
@@ -0,0 +1,250 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Datastore methods for operating on source IP allowlists.
+
+use crate::authz;
+use crate::context::OpContext;
+use crate::db::error::public_error_from_diesel;
+use crate::db::error::ErrorHandler;
+use crate::db::fixed_data::allowed_source_ips::ALLOWED_SOURCE_IPS_ID;
+use crate::db::DbConnection;
+use async_bb8_diesel::AsyncRunQueryDsl;
+use diesel::ExpressionMethods;
+use diesel::QueryDsl;
+use diesel::SelectableHelper;
+use nexus_db_model::schema::allowed_source_ip;
+use nexus_db_model::AllowedSourceIp;
+use omicron_common::api::external::AllowedSourceIps;
+use omicron_common::api::external::Error;
+use omicron_common::api::external::LookupType;
+use omicron_common::api::external::ResourceType;
+
+impl super::DataStore {
+    /// Fetch the list of allowed source IPs.
+    ///
+    /// This is currently effectively a singleton, since it is populated by RSS
+    /// and we only provide APIs in Nexus to update it, not create / delete.
+    pub async fn allowed_source_ips_view(
+        &self,
+        opctx: &OpContext,
+    ) -> Result<AllowedSourceIp, Error> {
+        let conn = self.pool_connection_authorized(opctx).await?;
+        opctx.authorize(authz::Action::Read, &authz::FLEET).await?;
+        allowed_source_ip::dsl::allowed_source_ip
+            .find(ALLOWED_SOURCE_IPS_ID)
+            .first_async::<AllowedSourceIp>(&*conn)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel(
+                    e,
+                    ErrorHandler::NotFoundByLookup(
```

Comment:
I think this is correct just want to think out loud - we don't insert an initial value for this, right? (Because the whole point is that it needs to be set by the operator during RSS.) That means if someone (presumably Nexus when deciding to send firewall rules?) calls this method before the RSS handoff has completed, they will get an error and be unable to do whatever they wanted to do. That seems right as long as it doesn't accidentally block RSS from completing somehow (which it doesn't seem like it should! but I could imagine being surprised that it does so given how tricky it is to follow the RSS flow)

Reply:
That's right, we do not insert an initial value here. That's done in the rack-initialization routine `DataStore::rack_set_initialized()`, where we call `DataStore::allowed_source_ips_upsert_on_connection()` to insert the row. Since we never read this value from Nexus except through plumbing the service firewall rules or in response to a public API call, I think this is reasonable to do and won't block that RSS process.